src/static.jl

```diff
@@ -366,11 +376,82 @@ IfElse.ifelse(::True, x, y) = x
 
 IfElse.ifelse(::False, x, y) = y
 
+"""
+    StaticSymbol
+
+A statically typed `Symbol`.
+"""
+struct StaticSymbol{s}
+    StaticSymbol{s}() where {s} = new{s::Symbol}()
+    StaticSymbol(s::Symbol) = new{s}()
+end
+
+Base.Symbol(::StaticSymbol{s}) where {s} = s::Symbol
+
+Base.show(io::IO, ::StaticSymbol{s}) where {s} = print(io, "static(:$s)")
+
+is_static(x) = is_static(typeof(x))
+is_static(::Type{T}) where {T<:StaticInt} = True()
+is_static(::Type{T}) where {T<:StaticBool} = True()
+is_static(::Type{T}) where {T<:StaticSymbol} = True()
+is_static(::Type{T}) where {T} = False()
+
+_tuple_static(::Type{T}, i) where {T} = is_static(_get_tuple(T, i))
+function is_static(::Type{T}) where {N,T<:Tuple{Vararg{Any,N}}}
+    return all(eachop(_tuple_static, T, nstatic(Val(N))))
+end
+
 """
     static(x)
 
-Returns a static form of `x`.
+Returns a static form of `x`. If `x` is already in a static form then `x` is returned. If
+there is no static alternative for `x` then an error is thrown.
+
+```julia
+julia> using ArrayInterface: static
+
+julia> static(1)
+static(1)
+
+julia> static(true)
+ArrayInterface.True()
+
+julia> static(:x)
+static(:x)
+
+```
 """
+function static(x)
+    if is_static(x) isa True
+        return x
+    else
+        _no_static_type(x)
+    end
+end
 static(x::Int) = StaticInt(x)
 static(x::Bool) = StaticBool(x)
+static(x::Symbol) = StaticSymbol(x)
+static(x::Tuple{Vararg{Any}}) = map(static, x)
```

Comment:
Perhaps we should try the aggressive constprop macro here, just for the sake of it?

I don't know the exact version number that added it, but if you want a safe bound:
```julia
help?> Base.@aggressive_constprop
  @aggressive_constprop ex
  @aggressive_constprop(ex)


  @aggressive_constprop requests more aggressive interprocedural constant propagation for the annotated function. For a method where the return type depends on the value of the arguments, this can yield improved inference results at the cost of additional compile time.

julia> VERSION
v"1.7.0-DEV.421"
```
So, for example (`@static` necessary to avoid `ERROR: LoadError: UndefVarError: @aggressive_constprop not defined`)
```julia
@static if VERSION >= v"1.7.0-DEV.421"
    Base.@aggressive_constprop static(x::Int) = StaticInt(x)
    Base.@aggressive_constprop static(x::Bool) = StaticBool(x)
    Base.@aggressive_constprop static(x::Symbol) = StaticSymbol(x)
    Base.@aggressive_constprop static(x::Tuple{Vararg{Any}}) = map(static, x)
else
    static(x::Int) = StaticInt(x)
    static(x::Bool) = StaticBool(x)
    static(x::Symbol) = StaticSymbol(x)
    static(x::Tuple{Vararg{Any}}) = map(static, x)
end
```
And similar everywhere else that the return type depends on values.